### PR TITLE
Fix throwing an exception if compilation-unit is not found

### LINKED ISSUE #42159: [Bug]: Getting error when getting the symbol for a class definition which is defined in a separate module
### Description

When I call `SemanticModel.symbol(ClassDefinitionNode)` I am getting a Bad Sad error. This is happening only when the class is defined in a separate module.

```console
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in
https://github.com/ballerina-platform/ballerina-lang/issues with the
log you get below and your sample code.

We thank you for helping make us better.

[2024-02-14 17:01:36,200] SEVERE {b7a.log.crash} - No value present 
java.util.NoSuchElementException: No value present
        at java.base/java.util.Optional.get(Optional.java:143)
        at io.ballerina.compiler.api.impl.BallerinaSemanticModel.getCompilationUnit(BallerinaSemanticModel.java:495)
        at io.ballerina.compiler.api.impl.BallerinaSemanticModel.symbol(BallerinaSemanticModel.java:181)
```

### Steps to Reproduce

1. Create a Ballerina project
2. Add a module in the project
3. Define a simple Ballerina class inside the new module
```ballerina
class Client {}
```
4. Try to get the symbol related to this ClassDefinitionNode using the semantic model

### Affected Version(s)

Ballerina SwanLake 2201.8.4

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by dulajdilshan**
## Purpose
$title

Fixes #42159

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/TestSourcesTest.java [hunk 1]
```diff
@@ -126,4 +133,72 @@ public Object[][] getExprPos() {
...
+            public void visit(ConstantDeclarationNode constantDeclarationNode) {
+                Optional<Symbol> foundSymbol = semanticModel.symbol(constantDeclarationNode);
+                assertTrue(foundSymbol.isEmpty());
+                constantDeclarationNode.initializer().accept(this);
+            }
+
+            @Override
+            public void visit(BasicLiteralNode basicLiteralNode) {
+                Optional<TypeSymbol> foundTypeSymbol = semanticModel.typeOf(basicLiteralNode);
+                assertTrue(foundTypeSymbol.isEmpty());
```
Comment (poorna2152) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/TestSourcesTest.java:200]: Can we merge these two lines?
```suggestion
                assertTrue(semanticModel.typeOf(basicLiteralNode).isEmpty());
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/test-project/modules/baz/baz.bal [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2024 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/test-project/modules/baz/baz.bal:1]:
```suggestion
// Copyright (c) 2024 WSO2 LLC. (http://www.wso2.org).
```

**Review by LakshanWeerasinghe — APPROVED** [review]
LGTM!

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java [hunk 3]
```diff
@@ -273,16 +289,25 @@ public List<Location> references(Document sourceDocument,
             return Collections.emptyList();
         }
         Location symbolLocation = symbolAtCursor.getPosition();
+
+        Optional<BLangCompilationUnit> compilationUnit = getCompilationUnit(targetDocument);
+        if (compilationUnit.isEmpty()) {
+            return Collections.emptyList();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:295]: codecov warnings?

Comment (dulajdilshan) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:295]: Fixed

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java [hunk 4]
```diff
@@ -166,8 +170,11 @@ public List<Symbol> visibleSymbols(Document sourceFile, LinePosition position, D
      */
     @Override
     public Optional<Symbol> symbol(Document sourceDocument, LinePosition position) {
-        BLangCompilationUnit compilationUnit = getCompilationUnit(sourceDocument);
-        return lookupSymbol(compilationUnit, position);
+        Optional<BLangCompilationUnit> compilationUnit = getCompilationUnit(sourceDocument);
+        if (compilationUnit.isEmpty()) {
+            return Optional.empty();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:175]: Can't you just return `compilationUnit`?

**Review by MaryamZi — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java (hunk 4 above)
Comment (dulajdilshan) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:175]: Here, the idea is to return `Optional<Symbol>` not `Optional<BLangCompilationUnit>`.

**Review by LakshanWeerasinghe — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java (hunk 4 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:175]: My bad, this is OK.

**Comment by MaryamZi** [comment]
Please check the build failures (in case they are relevant to this change).

**Comment by dulajdilshan** [comment]
> Please check the build failures (in case they are relevant to this change).

Already triggered

**Review by LakshanWeerasinghe — APPROVED** [review]

**Pull request merged**